lib/spack/spack/test/installer.py

```diff
@@ -252,6 +252,50 @@ def test_installer_str(install_mockery):
     assert "failed (0)" in istr
 
 
+def test_installer_prune_built_build_deps(install_mockery, monkeypatch, tmpdir):
+    r"""
+    Ensure that build dependencies of installed deps are pruned
+    from installer package queues.
+
+                (a)
+                / \
+               /   \
+             (d)   (b) <-- is installed already so we should
+               \   / \     prune (c) from this install since
+                \ /  (c)   it is *only* needed to build (b)
+                (e)
+
+    Thus since (b) is already installed our build_pq dag should
+    only include four packages. [(a), (b), (d), (e)]
+    """
+
+    @property
+    def _mock_installed(self):
+        return self.name in ["b"]
+
+    # Mock the installed property to say that (b) is installed
+    monkeypatch.setattr(spack.spec.Spec, "installed", _mock_installed)
+
+    # Create mock repository with packages (a), (b), (c), (d), and (e)
+    builder = spack.repo.MockRepositoryBuilder(tmpdir.mkdir("mock-repo"))
+
+    builder.add_package("a", dependencies=[("b", "build", None), ("d", "build", None)])
+    builder.add_package("b", dependencies=[("c", "build", None), ("e", "build", None)])
+    builder.add_package("c")
+    builder.add_package("d", dependencies=[("e", "build", None)])
+    builder.add_package("e")
+
+    with spack.repo.use_repositories(builder.root):
+        const_arg = installer_args(["a"], {})
+        installer = create_installer(const_arg)
+
+        installer._init_queue()
+
+        # Assert that (c) is not in the build_pq
+        for _, task in installer.build_pq:
+            assert "c-" not in task.pkg_id
```

Comment:
This needs to check not only that `c` isn't in the list, but also that all of the others are.

Maybe you could take `[task.pkg_id[0] for _, task in installer.build_pq]` and check that it is the list `["a", "b", "d", "e", "f"]`. (and maybe that gets easier to express if you make the one that should be excluded be last alphabetically, but maybe that's too fussy).

Reply:
Done!